docs/proposal/20210823-gpu-support.md

```diff
@@ -0,0 +1,197 @@
+
+### Goals
+
+1. To allow vGPU support for multi-node clusters.
+2. To allow GPU Direct support for multi-node clusters.
+3. To allow PCI passthrough support for single-node clusters.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - using vGPU
+
+Stacy works at an IT department of an organization where AI/ML models are run every hour and the entire staff wants to leverage GPU-supported nodes. VGPU's are able to provide close to bare-metal performance. Stacy needs a lot of sharing GPUs.
+vGPU provides a full-fledged cluster from which a GPU can be requested.
+
+#### Story 2 - GPU Direct implementation
+
+Tony is an Engineer at a big organization with multiple centers across the Globe. Tony needs a big GPU cluster to address the needs of his organization. So they intend to use GPU Direct. GPU direct is technically a pool of GPUs connected by a network card in a kind of peer-to-peer network. This is the latest technology and gives the advantage of shared GPUs from a pool. GPU  direct is the technology that helps in linking multiple GPUs. Tony wants to leverage this technology so that it is easier for him to manage resource allocation.
+
+#### Story 3 - PCI passthrough for single node Customer
+
+Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.
+
+### Requirements
+
+#### Functional Requirements
+
+- FR1: CAPV MUST support vGPUs.
+- FR2: CAPV MUST support GPUDirect.
+
+- FR3: CAPV MUST support PCI Passthrough.
+
+#### Non-Functional Requirements
+
+- NFR1: Unit tests MUST exist for all 3 supports
+- NFR2: e2e tests MUST exist for 3 supports
+
+### Implementation Details/Notes/Constraints
+
+#### Proposed Changes
+
+```go
+
+// VirtualMachineCloneSpec is information used to clone a virtual machine.
+type VirtualMachineCloneSpec struct {
+  // Vgpu is the Vgpu configuration for this machine's VM.
+ // +optional
+ Vgpu []VGPUSpec `json:"vgpu,omitempty"`
+ // Pci is the Pci configuration for this machine's VM.
+ // +optional
+ Pci []PCISpec `json:"pci,omitempty"`
+}
+
+//PCISpec defines virtual machine's PCI configuration
+type PCISpec struct {
+ // Devices is the list of pci devices used by the virtual machine.
+ PciDevices []PCIDeviceSpec `json:"pcidevices"`
+}
+
+//VGPUSpec defines virtual machine's VGPU configuration
+type VGPUSpec struct {
```

Comment:
I don't think we need this struct

Reply:
The VGPUSpec adds nothing apart from a wrapper around the VGPUDeviceSpec. Makes sense to remove this too. The NetworkSpec comprises of a bunch of other fields which are a part of the Spec, hence having the wrapper struct makes more sense in that case. And this approach applies to all the other structs above.

Reply:
Done!